ntex-bytes/src/buf/pool.rs

```diff
@@ -0,0 +1,269 @@
+{
+    type Backing = T::Backing;
+    type Owned = T::Owned;
+
+    fn put_back(&self, backing: Self::Backing) {
+        (&**self).put_back(backing);
+    }
+}
+
+impl<T> BufferSink for Arc<T>
+where
+    T: BufferSink,
+{
+    type Backing = T::Backing;
+    type Owned = T::Owned;
+
+    fn put_back(&self, backing: Self::Backing) {
+        (&**self).put_back(backing);
+    }
+}
+
+/// A [`BufferSource`] that always yields new `BytesMut` without blocking.
+#[derive(Clone, Copy, Debug)]
+pub struct BytesPool;
+
+impl BufferSource for BytesPool {
+    type Owned = BytesMut;
+    type Accumulator = BytesAccumulator;
+    type TakeFuture = future::Ready<Self::Owned>;
+
+    fn take(&self, len: usize) -> Self::TakeFuture {
+        future::ready(BytesMut::with_capacity(len))
+    }
+}
+
+impl BufferSink for BytesPool {
+    type Backing = ();
+    type Owned = <Self as BufferSource>::Owned;
+
+    fn put_back(&self, _backing: Self::Backing) {
+    }
+}
+
+/// A buffer that owns a particular range of its backing buffer.
+///
+/// An `Owned` tracks what part of itself has been filled with data.
+/// The filled region can be accessed with [`Owned::filled`].
+/// The unfilled region can be accessed with [`Owned::unfilled_mut`], and bytes can be moved from the start of this region
+/// into the end of the filled region with [`Owned::fill`].
+///
+/// An `Owned` can be subdivided into smaller `Owned`s with `split_to` that each own
+/// smaller splits of the backing buffer.
+///
+/// An `Owned` is not `Clone`. It can be converted to a [`Shared`] which is, via [`Owned::freeze`]
+pub trait Owned {
+    /// The type of shared buffers with the same backing buffer.
+    type Shared: Shared;
+
+    /// Try to append a `u8` to this buffer's filled region.
+    ///
+    /// Returns `None` if there is no space in the filled region to append this `u8`.
+    fn try_put_u8(&mut self, n: u8) -> Option<()> {
+        self.try_put_slice(&[n])
+    }
+
+    /// Try to append a `u16` in big-endian order to this buffer's filled region.
+    ///
+    /// Returns `None` if there is no space in the filled region to append this `u16`.
+    fn try_put_u16_be(&mut self, n: u16) -> Option<()> {
+        self.try_put_slice(&n.to_be_bytes())
+    }
+
+    /// Try to append a `u32` in big-endian order to this buffer's filled region.
+    ///
+    /// Returns `None` if there is no space in the filled region to append this `u32`.
+    fn try_put_u32_be(&mut self, n: u32) -> Option<()> {
+        self.try_put_slice(&n.to_be_bytes())
+    }
+
+    /// Try to append a slice in big-endian order to this buffer's filled region.
+    ///
+    /// Returns `None` if there is no space in the filled region to append this slice.
+    fn try_put_slice(&mut self, src: &[u8]) -> Option<()>;
+
+    /// Returns the filled region of this buffer.
+    fn filled(&self) -> &[u8];
+
+    /// Returns whether the filled region of this buffer is empty or not.
+    fn filled_is_empty(&self) -> bool {
```

Comment:
I don't think `filled_` prefix is necessary

Reply:
I'd rather be explicit about filled vs unfilled. If you had never seen the type before, you would not know if `is_empty` refers to the filled region (that you can read from) or unfilled region (that you can write to).

Reply:
but this would mean we use different names from known rust types like vec, bytes etc. all of them uses same terms and familiar

Reply:
`Vec` does not have a concept of "filled" and "unfilled" regions as a first-class concept. With an `Owned` you're equally interested in both the filled region (as a reader) and the unfilled region (as a writer). With `Vec` you're primarily interested in only the filled region, so it's natural to assume `is_empty()` etc refer to the filled region.

`BytesMut`'s names *are* ambiguous. It's not obvious from looking at the name `is_empty` or `len` whether it's referring to the filled region or unfilled region of the `BytesMut`. You have to read the docs to see it's referring to the filled region. I don't want to propagate this bad naming unless you insist.

Reply:
i do not agree. bytes is pretty consistent, len, is_empty is related to data, capacity is total allocations. i do not see much sense in filled_is_empty(), as there is no sense unfill_is_empty(). len, is_empty are related to actual data in storage. capacity is size of underlying storage and Vec, Bytes are pretty consistent and represent filled and unfilled storages.